feat: UIG-2709 - vl-datepicker-next - introductie component [zonder "visual-format"]

Goldflow opened the pull request:
[bamboo](https://www.milieuinfo.be/bamboo/browse/UIGOV-CUWC186)
[jira](https://www.milieuinfo.be/jira/browse/UIG-2709)

---

we adviseren vanaf nu zelf conversie van het datum formaat applicatief te doen
`visual-format` / `altFormat` wordt niet meer ondersteund

meer info: https://github.com/milieuinfo/uigov-web-components/pull/197

---

Er zijn een paar courante formaten die binnen DV / Departement Omgeving worden gebruik; o.m. het standaard formaat voor de datepicker `d.m.Y` (die resulteert in `21.02.2023`).

In de huidige implementatie, kunnen afnemers dus voor het gekozen formaat in het attribuut `pattern` ook een regex voorzien, er additioneel ook validatie is op het formaat.

Vraag me af of het wenselijk is, om afnemers hierbij te helpen:
- dat afnemers ofwel per (meest gebruikte) formaten een regex voorzien krijgen in documentatie (die ze dan zelf in het attribuut `pattern` kunnen steken)
- dat wij automagisch; al dan niet met een extra attribuut (bv. `auto-pattern`) zelf een RegEx voorzien voor de meest courante formaten (maar ik voel aan dat dit een stap te ver is)

tonyvandeneynde commented on apps/playground-native/src/app/app.element.ts:
```diff
@@ -1,143 +1,275 @@
-import { CascaderItem, TemplateFn, VlAccordionComponent, VlCascaderComponent, VlInfoTile } from '@domg-wc/components';
+import { VlCheckboxComponent } from '@domg-wc/components/next/form/checkbox';
```
Moet deze niet in registerWebComponents meegegeven worden?

Goldflow replied:
Verwijderd.

tonyvandeneynde commented on libs/components/src/next/form/checkbox/stories/vl-checkbox.stories-doc.mdx:
```diff
@@ -0,0 +1,45 @@
+
+Gebruik het `checkbox` component om de gebruiker de mogelijkheid te geven om 1 of meerdere dingen te selecteren in een lijst.<br/>
+
+Dit component is de volgende versie van het [vl-checkbox](?path=/story/components-next-form-checkbox--checkbox-default) component, we raden aan deze versie te gebruiken.<br />
```
Is het niet deze en de component?

Goldflow replied:
Aangepast

tonyvandeneynde commented on libs/components/src/next/form/datepicker/stories/vl-datepicker.stories-arg.ts:
```diff
@@ -0,0 +1,140 @@
+    selectedDate: {
+        name: 'selected-date',
+        description:
+            "Attribuut voor een vooringestelde datum conform het ingestelde formaat (bv. '03-10-2019') of 'today' voor vandaag.",
```
vooraf ingestelde

Goldflow replied:
had dit overgenomen van de oude datepicker docs; heb dit nu overal aangepast zoals voorgesteld

tonyvandeneynde commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,385 @@
+        let options = {
+            ...staticOptions,
+            ...this.getDynamicOptions(),
+        } as unknown as Options;
```
Hoeft hier niet eerst naar unknown gecast te worden. this.getDynamicOptions() geeft al Options terug. Er nog extra dingen aan toevoegen maakt voor Typescript niet uit.

tonyvandeneynde commented on libs/components/src/next/form/input-field/vl-input-field.component.ts:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    protected onInput(event: Event & { target: HTMLInputElement }): void {
+        this.value = event?.target?.value;
```
event is erboven getyped als iets dat niet null of undefined is en altijd een target property heeft. Dus ?. is hier niet nodig. -> event.target.value

Goldflow replied:
Aangepast

Goldflow committed:
feat: UIG-2709 - vl-datepicker-next - introductie component

Lit Playground voorbeeld toegevoegd.
altFormat / visual-format wordt niet meer ondersteund.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+            type: { type: String },
+            format: { type: String },
+            selectedDate: { attribute: 'selected-date', type: String },
+            minDate: { attribute: 'min-date', type: String },
```
Liefst eerst altijd type en daarna attribute

Goldflow replied:
Done.

ksvlaanderen commented on libs/components/src/next/form/datepicker/stories/vl-datepicker.stories-arg.ts:
```diff
@@ -0,0 +1,162 @@
+export const datepickerArgs = {
+    ...defaultArgs,
+    ...formControlArgs,
+    block: false,
```
Kan je hier je datepickerDefaultArgs gebruiken?

Goldflow replied:
Done!

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+        }
+    }
+
+    private getDatePicker() {
```
Kan je deze private methodes onder de lit lifecycle methods plaatsen?
Liefst altijd de lit lifecycle methods groeperenn

Goldflow replied:
Done!

ksvlaanderen commented on libs/components/src/next/form/datepicker/stories/vl-datepicker.stories-arg.ts:
```diff
@@ -0,0 +1,149 @@
+    },
+    type: {
+        name: 'type',
+        description: "Attribuut bepaalt het soort datepicker; range | time | date-time | '' (default)",
```
Overal 'Attribuut ...' in de descriptions verwijderen aub.

Goldflow replied:
Done.

ksvlaanderen commented on libs/components/src/next/form/datepicker/stories/vl-datepicker.stories-arg.ts:
```diff
@@ -0,0 +1,149 @@
+            defaultValue: { summary: datepickerArgs.maxTime },
+        },
+    },
+    amPm: {
```
Heeft de vorige versie van dit component ook ondersteuning voor amPm? Lijkt me overbodig voor de Vlaamse Overheid.

Goldflow replied:
Vorige had dit ook dus heb dit zo behouden, maar gezien dit in Vlaamse context gebruikt wordt lijkt me dit idd overbodig.
@kspeltix > OK voor het verwijderen van `amPm`-functionaliteit?

Goldflow replied:
Hebben uiteindelijk besloten dit te behouden.

ksvlaanderen commented on libs/components/src/next/form/datepicker/stories/vl-datepicker.stories-doc.mdx:
Dit is nog de oude manier om de story docs op te bouwen, kan je dit aanpassen?

ksvlaanderen commented on libs/components/src/next/form/datepicker/stories/vl-datepicker.stories-util.ts:
```diff
@@ -0,0 +1,19 @@
+
+export const formatEpoch = (epoch: number | string | symbol, format: string | symbol) => {
+    if (epoch && (typeof epoch === 'string' || typeof epoch === 'number')) {
+        // @ts-ignore
```
Kan je hier een reden voor de ts-ignore bij plaatsen?

Goldflow replied:
Verwijderd

ksvlaanderen commented on libs/components/src/next/form/datepicker/stories/vl-datepicker.stories-util.ts:
```diff
@@ -0,0 +1,19 @@
+
+export const formatTime = (time: string | symbol) => {
+    if (time && typeof time === 'string') {
+        return flatpickr.formatDate(new Date(`1970-01-01T${time}`), 'H:i');
```
Is er een betere manier om dit te doen ipv een fake datum te gebruiken?

Goldflow replied:
Ik bekijk het.
edit: op zich kon dit anders en kunnen die uren daar ook uit gehaald worden op een andere manier; 't is gwn voor deze functie maakt de datum niet uit, enkel de uur/minuten.

iig > functie verwijderd (werd niet gebruikt)

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.cy.ts:
```diff
@@ -0,0 +1,317 @@
+
+registerWebComponents([VlDatepickerComponent, VlErrorMessageComponent]);
+
+// function that returns the current date in the format Y-m-d
```
Nederlands aub

Goldflow replied:
Done.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.cy.ts:
```diff
@@ -0,0 +1,317 @@
+    const year = selectedYear ? selectedYear : date.getFullYear();
+    const month = selectedMonth ? selectedMonth : date.getMonth() + 1;
+    const day = selectedDay ? selectedDay : date.getDate();
+    // if day is less than 10, add a 0 in front of it
```
Nederlands aub

Goldflow replied:
Done

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.cy.ts:
Kan je wat meer accessible checks toevoegen in je testen? Het component in een bepaalde staat krijgen en dan de a11y testen.

Goldflow replied:
Heb enkele a11y testen toegevoegd.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+    type: '' as 'range' | 'time' | 'date-time',
+    value: '',
+    format: 'd.m.Y',
+    selectedDate: '',
```
Wat is het verschil tussen selectedDate en value?

Goldflow replied:
selectedDate verwijderd

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+    private pattern = DatepickerDefaults.pattern;
+
+    // Variables
+    instance?: Instance;
```
Hier zou ik `instance: Instance | null = null` van maken zoals we op andere plaatsen doen.

Goldflow replied:
Done.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+        super();
+
+        // Fix tot <-> tot-en-met
+        if (Dutch?.nl) {
```
Kan je dit in de connectedCallback doen?

Goldflow replied:
Done.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+
+        this.initializeComponent();
+
+        if (!this.isInShadowDOM()) {
```
Gaan we niet altijd gebruik maken van een Shadow DOM?

Goldflow replied:
Verwijderd.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+        };
+
+        return html`
+            <div class="vl-input-group" id="wrapper">
```
Is dit id nodig? Indien dit nodig is kan je het een zinvollere naam geven? datepicker-wrapper bv.

Goldflow replied:
Zinvollere naam gegeven.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+                    ?success=${this.success}
+                    ?readonly=${this.readonly}
+                    pattern=${this.pattern}
+                    aria-invalid=${this.error}
```
Is dit attribuut nodig?

Goldflow replied:
Echt nodig niet; onze vroegere form-validation.lib plaatste ook wel `aria-invalid` op form controls die in fout waren maar het is wel zo dat we dit tot nu toe nog niet hebben overgenomen.

Voorlopig verwijderd.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+                    ?required=${this.required}
+                    ?disabled=${this.disabled}
+                    ?error=${this.error}
+                    ?success=${this.success}
```
Niet nodig volgens mij.

Goldflow replied:
Gaat dit over success?

ksvlaanderen replied:
Inderdaad, het success attribuut bestaat niet op een input element.

ksvlaanderen replied:
Kan je deze nog fixen?

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+                <input
+                    id=${this.id}
+                    name=${this.name || this.id}
+                    type="text"
```
Moet dit niet type="date" zijn?

Goldflow replied:
Nee, want dan ga je in chromium browsers (& andere browsers die dit ondersteunen) een native datepicker gebruiken die op eerste zicht ook geen vrije tekstingave aanvaard. ( https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date )

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+                    @input=${this.onInput}
+                />
+                <button
+                    id="button"
```
Id weglaten of zinvollere naam geven aub

Goldflow replied:
done

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+                    type="button"
+                    class="vl-input-addon js-vl-datepicker-toggle  ${classMap(buttonClasses)}"
+                    ?disabled=${this.disabled}
+                    aria-label="toggle"
```
Zinvollere beschrijving geven of weglaten aun.

Goldflow replied:
done

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+                    class="vl-input-addon js-vl-datepicker-toggle  ${classMap(buttonClasses)}"
+                    ?disabled=${this.disabled}
+                    aria-label="toggle"
+                    @click="${() => {
```
Hier een aparte methode meegeven aub, anders kan die click listener niet correct verwijderd worden, je geeft nu telkens een nieuwe instantie van een methode mee.

Goldflow replied:
done

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+    }
+
+    connectedCallback() {
+        super.connectedCallback();
```
ConnectedCallback en disconnectedCallback boven de render plaatsen aub.

Goldflow replied:
done; connectedCallback ook verwijderd

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+        super.disconnectedCallback();
+
+        if (this.instance) {
+            this.instance.destroy();
```
this.instance?.destroy()

Goldflow replied:
Done

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+        return this.getRootNode() instanceof ShadowRoot;
+    }
+
+    onInput = (event: Event & { target: HTMLInputElement }) => {
```
Publieke methodes boven private methodes plaatsen aub.

Goldflow replied:
Done & private gemaakt

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+        }
+    };
+
+    handleDateChange = (date: string) => {
```
Kan deze private zijn?

Goldflow replied:
Done

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+        this.updateValue(value);
+    };
+
+    updateValue = (value: string) => {
```
Kan deze private zijn?

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+        this.setValue(value);
+    };
+
+    getValue() {
```
Kan deze private zijn?

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+    /**
+     * Geeft focus aan het datepicker input element.
+     */
+    focus() {
```
Kan deze private zijn?

Goldflow replied:
Deze kan niet private zijn gezien we transitief extenden van `HTMLElement` via de `FormControlMixin`.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,370 @@
+        return this.getRootNode() instanceof ShadowRoot;
+    }
+
+    onInput = (event: Event & { target: HTMLInputElement }) => {
```
Kan je een vl-input event dispatchen conform de andere form componenten?

Goldflow replied:
ik heb het geplaatst bij `updateValue()` zodat event ook wordt gethrowd bij het resetten, in het geval er een initial value wordt gezet & als de gebruiker in de `calendar`-popup een waarde kiest (gezien `onInput` enkel bij manuele ingave triggert)

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.uig-css.ts:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    .flatpickr-calendar--push-top {
+        margin-top: 14px;
```
Rem waarden gebruiken aub

Goldflow replied:
TLDR; deze CSS verwijderd zonder impact

Deze klasse is een overblijfsel van de originele variant waarbij die `visual-format` ook nog included was (en die zorgde voor een veel complexer component waar dan ook een heleboel spuug-en plakwerk voor nodig was, .o.a. dynamisch zetten van deze klasse). Deze is niet meer nodig - verwijderd.

ksvlaanderen commented on libs/components/src/next/form/datepicker/stories/vl-datepicker.stories.ts:
```diff
@@ -0,0 +1,76 @@
+            inlineStories: false,
+        },
+    },
+    decorators: [(story: () => unknown) => html` <div style="height: 400px;">${story()}</div>`],
```
Decorator weglaten en div met inline style in story voorzien.

Goldflow replied:
Done.

ksvlaanderen commented on libs/components/src/next/form/datepicker/stories/vl-datepicker.stories.ts:
```diff
@@ -0,0 +1,76 @@
+    parameters: {
+        docs: {
+            page: datepickerDocs,
+            inlineStories: false,
```
InlineStories weghalen

Goldflow replied:
Done.

ksvlaanderen commented on libs/components/src/next/form/datepicker/stories/vl-datepicker.stories-util.ts:
```diff
@@ -0,0 +1,10 @@
+import flatpickr from 'flatpickr';
+
+export const formatEpoch = (epoch: number | string | symbol, format: string | symbol) => {
```
Return type voorzien

Goldflow replied:
Done.

ksvlaanderen commented on libs/components/src/next/form/datepicker/stories/vl-datepicker.stories.ts:
```diff
@@ -0,0 +1,76 @@
+                label=${label}
+                value=${value}
+                error=${error}
+                success=${success}
```
?'s plaatsen

Goldflow replied:
Done.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,380 @@
+@customElement('vl-datepicker-next')
+export class VlDatepickerComponent extends FormControl {
+    // Properties
+    private value = DatepickerDefaults.value;
```
Zelfde volgorde hanteren als in de DatepickerDefaults aub

Goldflow replied:
Done.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,380 @@
+    }
+
+    private onInput = (event: Event & { target: HTMLInputElement }) => {
+        console.log('onInput', event);
```
Console.log verwijderen aub

Goldflow replied:
done

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,380 @@
+            }
+        }
+        let parsedDate;
+        try {
```
Dit is raar

Goldflow replied:
reworked

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,380 @@
+    };
+
+    private updateValue = (value: string) => {
+        this.value = value;
```
Refactor lifecycle of value variable

Goldflow replied:
done

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,380 @@
+            }
+        }
+
+        if (!this.initialValue) {
```
Dit zou ik in de connectedCallback plaatsen.

Goldflow replied:
Done.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,380 @@
+        return this.shadowRoot?.querySelector('.flatpickr-wrapper') as HTMLDivElement;
+    }
+
+    private getFlatpickrCalendar() {
```
Return types please

Goldflow replied:
done

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,380 @@
+        };
+    }
+
+    getOptions(): Options {
```
private

Goldflow replied:
done

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.cy.ts:
```diff
@@ -0,0 +1,317 @@
+    `);
+};
+
+describe('story vl-datepicker in form', () => {
```
component ipv story

Goldflow replied:
Done.

ksvlaanderen commented on apps/playground-lit/src/app/app.element.ts:
```diff
@@ -813,6 +883,7 @@ export class AppElement extends LitElement {
             this.firstNameRequired = false;
             this.lastNameRequired = false;
             this.interestsRequired = false;
+            this.birthdateRequired = false;
```
Kan je de birthdate value zelf ook nog resetten boven deze if?

Goldflow replied:
Done.

ksvlaanderen commented on apps/playground-lit/src/app/app.element.ts:
```diff
@@ -831,6 +903,7 @@ export class AppElement extends LitElement {
+            this.birthdateReadonly = false;
             this.birthplaceReadonly = false;
             this.hobbiesReadonly = false;
             this.ageReadonly = false;
```
Kan je onderaan in deze file de value ook afbeelden als deze gesubmit is?

Goldflow replied:
Zoals besproken; dit werkt zoals verwacht.

ksvlaanderen commented on libs/components/src/next/form/datepicker/stories/vl-datepicker.stories-util.ts:
```diff
@@ -0,0 +1,10 @@
+        const dateFormat = format && typeof format !== 'symbol' ? format : 'd.m.Y';
+        return flatpickr.formatDate(new Date(Number(epoch)), dateFormat);
+    } else {
+        return undefined;
```
Specifiek undefined teruggeven vind ik een beetje raar, je kan dit weglaten en dan gebeurt dit automatisch

Goldflow replied:
True, maar dan geeft hij in m'n IDE `TS7030: Not all code paths return a value.`; als ik return type aanpas van `string | undefined` naar `: string | void`, wordt het wel goedgekeurd. Heb het dan zo aangepast.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,338 @@
+    private pattern = DatepickerDefaults.pattern;
+
+    // Variables
+    instance: Instance | null = null;
```
Kan je deze private maken? Als de afnemers die nodig hebben kunnen we die later nog public maken.

Goldflow replied:
Hebben we idd afgesproken; aangepast.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,338 @@
+    connectedCallback() {
+        super.connectedCallback();
+
+        // Fix tot <-> tot-en-met
```
Mag deze comment weg?

Goldflow replied:
Heb deze comment hier expliciet ingelaten, gezien die comment ook in de vorige component erbij stond - blijkbaar is dit de enige uitzondering waarbij localisation moet gepatchd worden. Verwijderd.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,338 @@
+                    class="vl-input-addon js-vl-datepicker-toggle  ${classMap(buttonClasses)}"
+                    ?disabled=${this.disabled}
+                    aria-label="toggle calendar"
+                    @click="${this.toggleCalendar}"
```
Double quotes weglaten aub

Goldflow replied:
Done.

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,338 @@
+                <button
+                    id="toggle-calendar"
+                    type="button"
+                    class="vl-input-addon js-vl-datepicker-toggle  ${classMap(buttonClasses)}"
```
Kan je de `vl-input-addon js-vl-datepicker-toggle` classes mee in het buttonClasses object steken?

Goldflow replied:
Heb dit voor button & input zo gedaan, zodat die conditionele classes niet opnieuw geëvalueerd moeten worden bij elke render maar heb aangepast zoals voorgesteld gezien we hier tot nu toe geen onderscheid in maakten (en dit nooit iets zal uitmaken performance wise).

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,338 @@
+    }
+
+    /**
+     * Geeft focus aan het datepicker input element.
```
Beetje een nutteloze comment

Goldflow replied:
Verwijderd

ksvlaanderen commented on libs/components/src/next/form/datepicker/vl-datepicker.component.ts:
```diff
@@ -0,0 +1,338 @@
+
+        this.instance?.destroy();
+    }
+    protected render(): TemplateResult {
```
Blank line toevoegen aub
